drivers/media/i2c/mira220.c

```diff
@@ -0,0 +1,2481 @@
+	 printk(KERN_INFO "[MIRA220]: HIGH EXP Read reg %d, val = %llu.\n",
+		MIRA220_EXP_TIME_HI_REG_NEW, readval);
+	 ret = cci_write(mira220->regmap, MIRA220_EXP_TIME_LO_REG_NEW,
+		0xAA, NULL);
+	 usleep_range(1000000, 1000000+100);
+	 ret = cci_read(mira220->regmap,MIRA220_EXP_TIME_LO_REG_NEW, &readval, NULL );
+	 printk(KERN_INFO "[MIRA220]: LOW EXP Read reg %d, val = %llu.\n",
+		MIRA220_EXP_TIME_LO_REG_NEW, readval);
+	 */
+
+	/* Set default mode to max resolution */
+	mira220->mode = &supported_modes[0];
+
+	printk(KERN_INFO "[MIRA220]: Entering init controls function.\n");
+
+	ret = mira220_init_controls(mira220);
+	if (ret)
+		goto error_power_off;
+
+	/* Initialize subdev */
+	mira220->sd.internal_ops = &mira220_internal_ops;
+	mira220->sd.flags |= V4L2_SUBDEV_FL_HAS_DEVNODE |
+			     V4L2_SUBDEV_FL_HAS_EVENTS;
+	mira220->sd.entity.function = MEDIA_ENT_F_CAM_SENSOR;
+
+	/* Initialize source pads */
+	mira220->pad[IMAGE_PAD].flags = MEDIA_PAD_FL_SOURCE;
+	mira220->pad[METADATA_PAD].flags = MEDIA_PAD_FL_SOURCE;
+
+	printk(KERN_INFO "[MIRA220]: Entering set default format function.\n");
+
+	/* Initialize default format */
+	mira220_set_default_format(mira220);
+
+	printk(KERN_INFO "[MIRA220]: Entering pads init function.\n");
+
+	ret = media_entity_pads_init(&mira220->sd.entity, NUM_PADS,
+				     mira220->pad);
+	if (ret) {
+		dev_err(dev, "failed to init entity pads: %d\n", ret);
+		goto error_handler_free;
+	}
+
+	printk(KERN_INFO
+	       "[MIRA220]: Entering subdev sensor common function.\n");
+
+	ret = v4l2_async_register_subdev_sensor(&mira220->sd);
+	if (ret < 0) {
+		dev_err(dev, "failed to register sensor sub-device: %d\n", ret);
+		goto error_media_entity;
+	}
+
+	/* Enable runtime PM and turn off the device */
+	pm_runtime_set_active(dev);
+	pm_runtime_enable(dev);
+	pm_runtime_idle(dev);
+
+	return 0;
+
+error_media_entity:
+	media_entity_cleanup(&mira220->sd.entity);
+
+error_handler_free:
+	mira220_free_controls(mira220);
+
+error_power_off:
+	mira220_power_off(dev);
+
+	return ret;
+}
+
+static void mira220_remove(struct i2c_client *client)
+{
+	struct v4l2_subdev *sd = i2c_get_clientdata(client);
+	struct mira220 *mira220 = to_mira220(sd);
+
+	v4l2_async_unregister_subdev(sd);
+	media_entity_cleanup(&sd->entity);
+	mira220_free_controls(mira220);
+
+	pm_runtime_disable(&client->dev);
+	if (!pm_runtime_status_suspended(&client->dev))
+		mira220_power_off(&client->dev);
+	pm_runtime_set_suspended(&client->dev);
+}
+
+static const struct dev_pm_ops mira220_pm_ops = {
+	SET_SYSTEM_SLEEP_PM_OPS(mira220_suspend, mira220_resume)
+		SET_RUNTIME_PM_OPS(mira220_power_off, mira220_power_on, NULL)
+};
+
+static const struct of_device_id mira220_dt_ids[] = { { .compatible =
+								"ams,mira220" },
+						      { /* sentinel */ } };
+MODULE_DEVICE_TABLE(of, mira220_dt_ids);
+
+static const struct i2c_device_id mira220_ids[] = { { "mira220", 0 }, {} };
+MODULE_DEVICE_TABLE(i2c, mira220_ids);
```

Comment:
You shouldn't need I2C ids. This driver needs config data, so you can't load via sysfs.